api/listener.go

```diff
@@ -95,27 +96,32 @@ func (cl *chainListener) ReceiveBlock(blk *block.Block) error {
 }
 
 // AddResponder adds a new responder
-func (cl *chainListener) AddResponder(responder apitypes.Responder) (listenerID string, err error) {
-	if cl.streamMap.Count() >= cl.capacity {
-		err = errorCapacityReached
-		return
-	}
+func (cl *chainListener) NewStreamID() (listenerID string) {
 	i := 0
-	for ; i < idRetry; i++ {
+	for ; i < _idRetry; i++ {
 		listenerID = cl.idGenerator.newID()
 		if _, exist := cl.streamMap.Get(listenerID); !exist {
 			break
 		}
 	}
-	if i == idRetry {
-		err = errNoID
-		return
+	if i == _idRetry {
+		// An new id is assumed to be found, because comninations (2^62) is far larger than capacity
+		log.L().Panic("No peer id available", zap.Error(errNoID))
+		panic(errNoID)
 	}
-	responder.SetID(listenerID)
-	cl.streamMap.Set(listenerID, responder)
 	return
 }
 
+// AddResponder adds a new responder
+func (cl *chainListener) AddResponder(responder apitypes.Responder) error {
+	if _, exist := cl.streamMap.Get(responder.ID()); !exist &&
+		cl.streamMap.Count() >= cl.capacity {
+		return errorCapacityReached
+	}
```

Comment:
feels it is logically clear (and safer) not to combine `!exist` and `Count() >=` together, something like
```
if exist {
    // since newID() generates a unique ID each time, this should not happen
    return errCollision or panic()
}
if Count() >= cl.capacity {
    return errorCapacityReached
}
```

Reply:
otherwise the below code will overwrite an existing responder
```
r := NewGRPCBlockListener(
		svr.coreService.NewStreamID(),
		func(resp interface{}) error {
			return stream.Send(resp.(*iotexapi.StreamBlocksResponse))
		},
		errChan,
	)
cl.AddResponder(r)
r1 := NewGRPCBlockListener(r.ID() ...)
cl.AddResponder(r1) // overwrites existing r
```

Reply:
> otherwise the below code will overwrite an existing responder
> 
> ```
> r := NewGRPCBlockListener(
> 		svr.coreService.NewStreamID(),
> 		func(resp interface{}) error {
> 			return stream.Send(resp.(*iotexapi.StreamBlocksResponse))
> 		},
> 		errChan,
> 	)
> cl.AddResponder(r)
> r1 := NewGRPCBlockListener(r.ID() ...)
> cl.AddResponder(r1) // overwrites existing r
> ```

Yes, this is my thought of this func.

Reply:
i don't get it, what's the use-case of using a new responder to overwrite/replace existing one?

Reply:
ok. the logic is reverted